ndcube/ndcube.py

```diff
@@ -1,29 +1,123 @@
 import numpy as np
-import matplotlib.pyplot as plt
 import sunpy.map
-from sunpy.map.mapcube import MapCube
 
 from ndcube import cube_utils
 from ndcube import wcs_util
-from ndcube import DimensionPair, SequenceDimensionPair
-from ndcube.visualization import animation as ani
-from ndcube.plot_helpers import _plot_1D_cube, _plot_2D_cube, _plot_3D_cube
+from ndcube.mixins import NDCubeSlicingMixin, NDCubePlotMixin
+from ndcube import DimensionPair
 
 
-__all__ = ['NDCube', 'NDCubeOrdered']
+__all__ = ['NDCubeBase', 'NDCube', 'NDCubeOrdered']
 
 
-class NDCube(astropy.nddata.NDData):
+class NDCubeMetaClass(abc.ABCMeta, InheritDocstrings):
+    """
+    A metaclass that combines `abc.ABCMeta` and `~astropy.utils.misc.InheritDocstrings`.
+    """
+
+
+class NDCubeBase(astropy.nddata.NDData, metaclass=NDCubeMetaClass):
+
+    @abc.abstractmethod
+    def pixel_to_world(self, quantity_axis_list, origin=0):
+        """
+        Convert a pixel coordinate to a data (world) coordinate by using
+        `~astropy.wcs.WCS.all_pix2world`.
+
+        Parameters
+        ----------
+        quantity_axis_list : `list`
+            A list of `~astropy.units.Quantity` with unit as pixel `pix`.
+
+        origin : `int`.
+            Origin of the top-left corner. i.e. count from 0 or 1.
+            Normally, origin should be 0 when passing numpy indices, or 1 if
+            passing values from FITS header or map attributes.
+            See `~astropy.wcs.WCS.wcs_pix2world` for more information.
+            Default is 0.
+
+        Returns
+        -------
+
+        coord : `list`
+            A list of arrays containing the output coordinates
+            reverse of the wcs axis order.
+        """
+
+    @abc.abstractmethod
+    def world_to_pixel(self, quantity_axis_list, origin=0):
+        """
+        Convert a world coordinate to a data (pixel) coordinate by using
+        `~astropy.wcs.WCS.all_world2pix`.
+
+        Parameters
+        ----------
+        quantity_axis_list : `list`
+            A list of `~astropy.units.Quantity`.
+
+        origin : `int`
+            Origin of the top-left corner. i.e. count from 0 or 1.
+            Normally, origin should be 0 when passing numpy indices, or 1 if
+            passing values from FITS header or map attributes.
+            See `~astropy.wcs.WCS.wcs_world2pix` for more information.
+            Default is 0.
+
+        Returns
+        -------
+
+        coord : `list`
+            A list of arrays containing the output coordinates
+            reverse of the wcs axis order.
+        """
+
+    @abc.abstractproperty
+    def dimensions(self):
+        """
+        Returns a named tuple with two attributes: 'shape' gives the shape
+        of the data dimensions; 'axis_types' gives the WCS axis type of each dimension,
+        e.g. WAVE or HPLT-TAN for wavelength of helioprojected latitude.
+        """
+
+    @abc.abstractproperty
+    def extra_coords(self):
+        """
```

Comment:
@DanRyanIrish can you suggest a docstring for this?

Reply:
How about something like:
```
"""
Returns extra coordinates not included in the NDCube's WCS object.

Returns
-------
result: `dict`
    Dictionary of extra coords where each key is the name of an extra
    coordinate supplied by user during instantiation of the NDCube.
    The value of each key is itself a dictionary with the following keys:
    'axis': `int`
        The number of the data axis to which the extra coordinate corresponds.
    'value': `astropy.units.Quantity` or array-like
        The value of the extra coordinate at each pixel/array element along the
        corresponding axis (given by the 'axis' key, above).  Note this means
        that the length of 'value' must be equal to the length of the data axis
        to which is corresponds.
        
        """
```
?